02-intermediate-front-end/events-and-callbacks/readme.md

```diff
@@ -33,77 +76,592 @@ button.addEventListener("click", handleClickEvent)
+(function(){
+  var x;
+  for(x = 2; x < 10; x++){
+    console.log(x + " black spiders! Ah ah ahh!")
+  }
+  console.log("I lahv counting! Ah ah ahh!")
+}());
+```
+
+This seems kind of pointless, right? The advantage is that it keeps variables from cluttering up your browser. When Javascript comes to the end of a function, it "forgets" all of the data it generated inside that function to free up memory. Keeping everything inside functions and IIFEs can have a significant impact on the performance of your website since it takes less memory to run.
+
+## Asynchronicity
+
+Give your internal monologue code a "farewell" message that runs at the end:
+
+```js
+(function averageMorning(first, then, lastly){
+  console.log("This is my internal monologue on a normal weekday")
+  first();
+  then();
+  setTimeout(lastly, 5000);
+  console.log("...and that's my day!")
+}(haveBreakfast, commute, workAtGA));
+```
+
+Wait, what? The farewell came before the "at GA" message!
+
+With everything else we've seen, Javascript executes one line of code, then when it's done, executes the next line of code. This is called being **synchronous**.
+
+However, some operations in Javascript are **asynchronous**, meaning Javascript goes on to the next line of code without waiting for the previous line to complete.
+
+This is limited mostly to timing functions, and operations where Javascript is loading data from some other website.
 
-Let’s start with a working solution and improve it.
+### Why doesn't Javascript wait for these operations to complete before going to the next line of code?
 
-    $ cd ~/wdi/exercises/color-scheme-switcher
-    $ git checkout origin/solution
+Because otherwise the webpage would just "hang" until the operation completes. The browser can't do anything while Javascript is actively running. We've seen this when we've encountered infinite loops. Asyncrhonicity is a way of preventing the computer from freezing.
 
-1. Replace inline styles with class manipulation.
-2. Use the same function for all event handlers  
-3. Use one event listener
+This risk is potentially the greatest when Javascript is making requests to other webpages. There's no way of knowing how long the request will take to complete. It could be near-instant, but if the target server is having a bad day, it could take who-knows-how-long. You don't want the operability of your computer to be at the mercy of some random computer somewhere else.
 
-### 1. Classes instead of inline styles
+In the internal monologue app we made, anything we want to be sure happens **after** those 5 seconds of commuting should go inside the `lastly` callback. This way, we can be certain that it will run only when the 5 seconds are up.
 
-Manipulating classes with JavaScript instead of manipulating inline styles allows us to separate our concerns.
-All style-related things will be in a css file (see how easy it is to add a new theme at the end!)
+-----
 
 ### 2. One Event Handler to Rule them all
 
```

Comment:
OVerall really great. I think I want to eliminate some of the depth in this lesson plan. And really give them a better foundation for just the click event even. Small admin thing, I think it'd be nice whenever using event as an argument to write out the word `event` instead of `e` to be more clear. Let me know what you think.